File: ui/GameNetworkManager.cpp
#include "GameNetworkManager.h"
#include <iostream>
#include <cstring>

GameNetworkManager::GameNetworkManager() {}
GameNetworkManager::~GameNetworkManager() { disconnect(); }

bool GameNetworkManager::connect(const std::string& host, unsigned short port) {
    return client_.connect(host, port);
}

void GameNetworkManager::disconnect() {
    client_.disconnect();
}

bool GameNetworkManager::isConnected() const {
    return client_.isConnected();
}

void GameNetworkManager::joinRoom(const std::string& roomId,
                                  OnlineMode mode, TimePreset timePreset) {
    client_.sendJoin(roomId,
                     static_cast<int>(mode),
                     static_cast<int>(timePreset));
}

void GameNetworkManager::sendReady() {
    client_.sendReady();
}

void GameNetworkManager::sendMove(int boardX, int boardY,
                                  const std::vector<std::vector<int>>& shape) {
    client_.sendMove(boardX, boardY, shape);
}

// ── Packet parsing ────────────────────────────────────────────────────────────

void GameNetworkManager::pollEvents(const NetworkCallbacks& callbacks) {
    NetPacket pkt;
    while (client_.pollPacket(pkt)) {
        handlePacket(pkt, callbacks);
    }
    // Only fire disconnect if we were previously connected and now lost it
    if (wasConnected_ && !client_.isConnected()) {
        wasConnected_ = false;
        if (callbacks.onDisconnected)
            callbacks.onDisconnected();
    }
    if (client_.isConnected()) wasConnected_ = true;
}

void GameNetworkManager::handlePacket(const NetPacket& pkt,
                                      const NetworkCallbacks& cb) {
    const auto& p = pkt.payload;

    switch (pkt.type) {

    case MsgType::ROOM_JOINED:
        // payload: [1 playerId]
        if (p.size() >= 1) {
            state_.myPlayerId = p[0];
            if (cb.onRoomJoined) cb.onRoomJoined(state_.myPlayerId);
        }
        break;

    case MsgType::GAME_START:
        // payload: [1 mode][2 timeLimitSec LE]
        if (p.size() >= 3) {
            OnlineMode mode = static_cast<OnlineMode>(p[0]);
            int timeSec = static_cast<int>(p[1]) | (static_cast<int>(p[2]) << 8);
            state_.myTimeMs       = timeSec * 1000;
            state_.opponentTimeMs = timeSec * 1000;
            if (cb.onGameStart) cb.onGameStart(mode, timeSec);
        }
        break;

    case MsgType::GAME_STATE:
        // payload: [64 bytes opponent grid row-major]
        if (p.size() >= Board::SIZE * Board::SIZE) {
            for (int y = 0; y < Board::SIZE; ++y)
                for (int x = 0; x < Board::SIZE; ++x)
                    state_.opponentGrid[x][y] = p[y * Board::SIZE + x];
            if (cb.onStateUpdate) cb.onStateUpdate(state_);
        }
        break;

    case MsgType::SCORE_UPDATE:
        // payload: [1 playerId][4 score LE][1 combo]
        if (p.size() >= 6) {
            int pid   = p[0];
            int score = p[1] | (p[2] << 8) | (p[3] << 16) | (p[4] << 24);
            int combo = p[5];
            if (pid == state_.myPlayerId) {
                state_.myScore = score;
                state_.myCombo = combo;
            } else {
                state_.opponentScore = score;
                state_.opponentCombo = combo;
            }
            if (cb.onStateUpdate) cb.onStateUpdate(state_);
        }
        break;

    case MsgType::TIME_UPDATE:
        // payload: [4 p0TimeMs LE][4 p1TimeMs LE]
        if (p.size() >= 8) {
            int t0 = p[0] | (p[1]<<8) | (p[2]<<16) | (p[3]<<24);
            int t1 = p[4] | (p[5]<<8) | (p[6]<<16) | (p[7]<<24);
            if (state_.myPlayerId == 0) {
                state_.myTimeMs       = t0;
                state_.opponentTimeMs = t1;
            } else {
                state_.myTimeMs       = t1;
                state_.opponentTimeMs = t0;
            }
            if (cb.onStateUpdate) cb.onStateUpdate(state_);
        }
        break;

    case MsgType::YOUR_TURN:
        state_.myTurn = true;
        if (cb.onYourTurn) cb.onYourTurn();
        break;

    case MsgType::GAME_OVER:
        // payload: [1 winnerId]
        if (p.size() >= 1) {
            state_.gameOver = true;
            state_.winnerId = p[0];
            if (cb.onGameOver) cb.onGameOver(state_.winnerId);
        }
        break;

    case MsgType::PONG:
        break;

    case MsgType::ERROR_MSG:
        std::cerr << "[Net] Server error\n";
        break;

    default:
        std::cerr << "[Net] Unknown packet type: "
                  << static_cast<int>(pkt.type) << "\n";
        break;
    }
}

File: UnityPort/Assets/Scripts/Network/GameNetworkManager.cs
using System;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace BlockBlast.Network
{
    using BlockBlast.Core;

    public enum OnlineMode
    {
        REALTIME = 0,
        TURN_BASED = 1
    }

    public enum TimePreset
    {
        FAST = 60,
        NORMAL = 180,
        LONG = 300
    }

    public class OnlineGameState
    {
        public int myScore = 0;
        public int opponentScore = 0;
        public int myCombo = 0;
        public int opponentCombo = 0;
        public int myTimeMs = 0;
        public int opponentTimeMs = 0;
        public bool myTurn = false;
        public bool gameOver = false;
        public int winnerId = -1;
        public int myPlayerId = -1;

        public int[,] opponentGrid = new int[Board.SIZE, Board.SIZE];
    }

    public class GameNetworkManager : MonoBehaviour
    {
        private NetworkClient client = new NetworkClient();
        private OnlineGameState state = new OnlineGameState();
        private bool wasConnected = false;

        public OnlineGameState State => state;
        public bool IsConnected => client.IsConnected;

        // --- Event Delegates ---
        public event Action<int> OnRoomJoined;
        public event Action<OnlineMode, int> OnGameStart;
        public event Action<OnlineGameState> OnStateUpdate;
        public event Action<int> OnGameOver;
        public event Action OnYourTurn;
        public event Action OnDisconnected;

        private void Update()
        {
            PollEvents();
        }

        private void OnDestroy()
        {
            Disconnect();
        }

        public async Task<bool> ConnectAsync(string host, int port)
        {
            state = new OnlineGameState();
            wasConnected = false;
            return await client.ConnectAsync(host, port);
        }

        public void Disconnect()
        {
            if (client.IsConnected)
            {
                client.Disconnect();
            }
        }

        public void JoinRoom(string roomId, OnlineMode mode, TimePreset timePreset)
        {
            if (client.IsConnected)
            {
                client.SendJoin(roomId, (int)mode, (int)timePreset);
            }
        }

        public void SendReady()
        {
            if (client.IsConnected) client.SendReady();
        }

        public void SendMove(int boardX, int boardY, int[,] shape)
        {
            if (client.IsConnected)
            {
                client.SendMove(boardX, boardY, shape);
                if (state.myTurn)
                {
                    state.myTurn = false; // Optimistic turn clearing
                }
            }
        }

        public void PollEvents()
        {
            bool isConn = client.IsConnected;
            if (isConn != wasConnected)
            {
                wasConnected = isConn;
                if (!isConn)
                {
                    OnDisconnected?.Invoke();
                }
            }

            if (!isConn) return;

            while (client.PollPacket(out NetPacket pkt))
            {
                HandlePacket(pkt);
            }
        }

        private void HandlePacket(NetPacket pkt)
        {
            switch (pkt.type)
            {
                case MsgType.ROOM_JOINED:
                    if (pkt.payload.Length >= 1)
                    {
                        state.myPlayerId = pkt.payload[0];
                        OnRoomJoined?.Invoke(state.myPlayerId);
                    }
                    break;

                case MsgType.GAME_START:
                    if (pkt.payload.Length >= 2)
                    {
                        OnlineMode mode = (OnlineMode)pkt.payload[0];
                        int timeSec = pkt.payload[1];
                        state.gameOver = false;
                        state.winnerId = -1;
                        state.myScore = 0;
                        state.opponentScore = 0;
                        state.myCombo = 0;
                        state.opponentCombo = 0;
                        state.myTurn = (mode == OnlineMode.TURN_BASED && state.myPlayerId == 0);

                        OnGameStart?.Invoke(mode, timeSec);
                    }
                    break;

                case MsgType.GAME_STATE:
                    // Payload is 64 bytes row-major grid representation of opponent board
                    if (pkt.payload.Length >= Board.SIZE * Board.SIZE)
                    {
                        for (int y = 0; y < Board.SIZE; y++)
                        {
                            for (int x = 0; x < Board.SIZE; x++)
                            {
                                // Grid index col-major (matching Board.cs structure)
                                state.opponentGrid[x, y] = pkt.payload[y * Board.SIZE + x];
                            }
                        }
                        OnStateUpdate?.Invoke(state);
                    }
                    break;

                case MsgType.SCORE_UPDATE:
                    // payload: [1 playerId][4 bytes score][1 byte combo]
                    if (pkt.payload.Length >= 6)
                    {
                        int pid = pkt.payload[0];
                        int score = BitConverter.ToInt32(pkt.payload, 1);
                        int combo = pkt.payload[5];

                        if (pid == state.myPlayerId)
                        {
                            state.myScore = score;
                            state.myCombo = combo;
                        }
                        else
                        {
                            state.opponentScore = score;
                            state.opponentCombo = combo;
                        }
                        OnStateUpdate?.Invoke(state);
                    }
                    break;

                case MsgType.TIME_UPDATE:
                    // payload: [4 bytes myTime ms][4 bytes oppTime ms]
                    if (pkt.payload.Length >= 8)
                    {
                        int t1 = BitConverter.ToInt32(pkt.payload, 0);
                        int t2 = BitConverter.ToInt32(pkt.payload, 4);
                        state.myTimeMs = t1;
                        state.opponentTimeMs = t2;
                        OnStateUpdate?.Invoke(state);
                    }
                    break;

                case MsgType.YOUR_TURN:
                    state.myTurn = true;
                    OnYourTurn?.Invoke();
                    break;

                case MsgType.GAME_OVER:
                    if (pkt.payload.Length >= 1)
                    {
                        state.gameOver = true;
                        state.winnerId = pkt.payload[0];
                        OnGameOver?.Invoke(state.winnerId);
                    }
                    break;

                case MsgType.ERROR_MSG:
                    string errMsg = Encoding.ASCII.GetString(pkt.payload);
                    Debug.LogWarning("[NetworkManager] Server error message: " + errMsg);
                    break;

                case MsgType.PONG:
                    // Log keep-alive response
                    break;
            }
        }
    }
}
